### PR TITLE
Include inner flow

### PR DESCRIPTION
**Pull request by Baraujo25**

### CODE REVIEW

File: src/main/java/us/abstracta/wiresham/Flow.java [hunk 1]
```diff
@@ -144,9 +169,65 @@ public static Flow fromPcap(File file, String serverAddress, String filter) thro
   }
 
   public static Flow fromYml(File ymlFile) throws FileNotFoundException {
-    List<PacketStep> packets = new Yaml(buildYamlConstructor())
-        .load(new FileInputStream(ymlFile));
-    return new Flow(packets);
+    List<Flow> flows = new ArrayList<>();
+    new Yaml(buildYamlConstructor())
```
Comment (rabelenda) [inline src/main/java/us/abstracta/wiresham/Flow.java]: An alternative can be to use streams with something like `StreamSupport.stream(iterable, false)`

Comment (rabelenda) [reply src/main/java/us/abstracta/wiresham/Flow.java:173]: And use flatmap for parseFlowsFromYml?

File: src/main/java/us/abstracta/wiresham/Flow.java [hunk 2]
```diff
@@ -144,9 +169,65 @@ public static Flow fromPcap(File file, String serverAddress, String filter) thro
...
+        .iterator()
+        .forEachRemaining(f -> flows.add((Flow) f));
+    parseFlowsFromYml(flows);
+    return flows.stream().filter(f -> f.getId() == null).findFirst().orElse(flows.get(0));
+  }
+
+  private static void parseFlowsFromYml(List<Flow> flows) {
+    for (Flow flow : flows) {
+      List<PacketStep> steps = new ArrayList<>(flow.steps);
+      Optional<IncludePacketStep> includePacket = getIncludePacket(steps);
```
Comment (rabelenda) [inline src/main/java/us/abstracta/wiresham/Flow.java:184]: are you expecting to only have one include per flow?

File: src/main/java/us/abstracta/wiresham/Flow.java [hunk 3]
```diff
@@ -144,9 +169,65 @@ public static Flow fromPcap(File file, String serverAddress, String filter) thro
...
+    List<Flow> flows = new ArrayList<>();
+    new Yaml(buildYamlConstructor())
+        .loadAll(new FileInputStream(ymlFile))
+        .iterator()
+        .forEachRemaining(f -> flows.add((Flow) f));
+    parseFlowsFromYml(flows);
+    return flows.stream().filter(f -> f.getId() == null).findFirst().orElse(flows.get(0));
+  }
+
+  private static void parseFlowsFromYml(List<Flow> flows) {
```
Comment (rabelenda) [inline src/main/java/us/abstracta/wiresham/Flow.java]: maybe a recursive logic simplifies this?

something like:

```java
return solveSteps(flows.get(0), flows);

solveSteps(flow, flows) {
  ret = []
  lastPort = defaultPort
  for ( step : flow.steps) {
    if (step instanceOf IncludePacketStep) {
      ret.addAll(solveSteps(findFlow(includeStep.id, flows), flows))
    } else {
      if step.port != null {
        lastPort = step.port
      } else {
        step.port = lastPort
      }
      ret.add(step)
    }
  }
  return ret
}

findFlow(id, flows) {
  flows.stream()
    .filter(includeStep.id.equals(it.id))
    .findFirst()
    .orElseGet(() -> flows.get(Integer.parseInt(id)))
}
```

File: src/main/java/us/abstracta/wiresham/Flow.java [hunk 4]
```diff
@@ -144,9 +169,65 @@ public static Flow fromPcap(File file, String serverAddress, String filter) thro
...
+        steps.addAll(includeIndex, getStepsById(flows, includePacket.get().getId()));
+        steps.remove(includePacket.get());
+        includePacket = getIncludePacket(steps);
+      }
+      flow.setSteps(steps);
+    }
+  }
+
+  private static List<PacketStep> getStepsById(List<Flow> flows, String id) {
+    return flows.stream().filter(f -> f.getId() != null)
```
Comment (rabelenda) [inline src/main/java/us/abstracta/wiresham/Flow.java]: if you use `id.equals(f.getId())` do you require 2 filters?

File: src/main/java/us/abstracta/wiresham/IncludePacketStep.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  // Constructor left for serialization/deserialization proposes 
+  public IncludePacketStep() {
+  }
+
+  // Constructor left for serialization/deserialization proposes 
+  public IncludePacketStep(String id) {
+    this.id = id;
+  }
+
+  @Override
```
Comment (rabelenda) [inline src/main/java/us/abstracta/wiresham/IncludePacketStep.java]: usually getters and setters are placed before other methods

File: src/main/java/us/abstracta/wiresham/PacketStep.java [hunk 6]
```diff
@@ -12,7 +12,7 @@ public abstract class PacketStep {
   public int port;
   protected Packet data;
  
-  protected PacketStep() {
+  public PacketStep() {
```
Comment (rabelenda) [inline src/main/java/us/abstracta/wiresham/PacketStep.java]: any particular reason for this change?

File: src/main/java/us/abstracta/wiresham/Flow.java (hunk 3 above)
Comment (Baraujo25) [reply src/main/java/us/abstracta/wiresham/Flow.java]: I did not think about that when implementing it. But know makes a lot of sense to use recursion over here.
Thanks for this satisfying contribution!!

**Review by rabelendameli — APPROVED** [review]

**Pull request merged**